api/transport/websocket/websocket.go

```diff
@@ -0,0 +1,162 @@
+package websocket
+
+import (
+	"errors"
+	"github.com/gorilla/websocket"
+	uuid "github.com/satori/go.uuid"
+	"github.com/spaceuptech/space-api-go/api/config"
+	"github.com/spaceuptech/space-api-go/api/model"
+	"log"
+	"sync"
+	"time"
+)
+
+const ServiceRegister string = "service-register"
+const ServiceRequest string = "service-request"
+
+type websocketOptions struct {
+	projectId string
+	token     string
+}
+
+type Socket struct {
+	url                  string
+	isConnected          bool
+	connectedOnce        bool
+	options              websocketOptions
+	pendingMsg           []model.WebsocketMessage
+	Socket               *websocket.Conn
+	SendMessage          chan model.WebsocketMessage
+	registerCallbackMap  map[string]func(data interface{})
+	onReconnectCallbacks []func()
+	mux                  sync.RWMutex
+}
+
+func Init(url string, config *config.Config) *Socket {
+	url = "ws://" + url + "/v1/api/" + config.Project + "/socket/json"
+	if config.IsSecure {
+		url = "wss://" + url + "/v1/api/" + config.Project + "/socket/json"
+	}
+	return &Socket{
+		url:                 url,
+		options:             websocketOptions{projectId: config.Project, token: config.Token},
+		registerCallbackMap: map[string]func(data interface{}){},
+		pendingMsg:          []model.WebsocketMessage{},
+		mux:                 sync.RWMutex{},
+	}
+}
+
+func (s *Socket) connect() error {
+	conn, _, err := websocket.DefaultDialer.Dial(s.url, nil)
+	if err != nil {
+		log.Println(err)
+		return err
+	}
+
+	s.setSocket(conn)
+	s.setConnected(true)
+
+	if s.getConnectedOnce() {
+		for _,value := range s.onReconnectCallbacks {
+			value()
+		}
+	}
+
+	if len(s.pendingMsg) != 0 {
+		for _, payload := range s.pendingMsg {
+			s.Socket.WriteJSON(payload)
+		}
+		s.pendingMsg = []model.WebsocketMessage{}
+	}
+
+	writeMessage := make(chan model.WebsocketMessage)
```

Comment:
Make this inside the Init method. You dont need to close this channel at all

Reply:
done